apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerException.cs

```diff
@@ -0,0 +1,173 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Linq;
+using Google.Cloud.Spanner.V1.Logging;
+using Grpc.Core;
+
+namespace Google.Cloud.Spanner
+{
+    /// <summary>
+    /// Represents an error communicating with the Spanner database.
+    /// </summary>
+    public sealed class SpannerException : Exception
+    {
+        /// <summary>
+        /// This class is a thin conversion around a grpc exception, with the additional
+        /// information of whether the operation is retryable based on the resulting error.
+        /// </summary>
+        /// <param name="code"></param>
+        /// <param name="innerException"></param>
+        internal SpannerException(ErrorCode code, RpcException innerException)
+            : base(GetMessageFromErrorCode(code), innerException)
+        {
+            Logger.LogPerformanceCounterFn("SpannerException.Count", x => x + 1);
+            ErrorCode = code;
+        }
+
+        internal SpannerException(ErrorCode code, string message)
+            : base(message)
+        {
+            Logger.LogPerformanceCounterFn("SpannerException.Count", x => x + 1);
+            ErrorCode = code;
+        }
+
+        internal SpannerException(RpcException innerException)
+            : this(ConvertFromStatusCode(innerException.Status.StatusCode), innerException)
+        {
+        }
+
+        /// <summary>
+        /// </summary>
+        public ErrorCode ErrorCode { get; }
+
+        /// <summary>
+        /// </summary>
+        public bool IsRetryable
+        {
+            get
+            {
+                switch (ErrorCode)
+                {
+                    case ErrorCode.DeadlineExceeded:
+                    case ErrorCode.Aborted:
+                    case ErrorCode.Unavailable:
+                        return true;
+                    default:
+                        return false;
+                }
+            }
+        }
+
+        internal static bool TryTranslateRpcException(Exception possibleRpcException,
+            out SpannerException spannerException)
+        {
+            spannerException = null;
+            var rpcException = possibleRpcException as RpcException;
+            if (rpcException != null)
+                spannerException = new SpannerException(rpcException);
+            var aggregateException = possibleRpcException as AggregateException;
+            if (aggregateException?.InnerExceptions != null)
+                spannerException = (SpannerException) aggregateException.InnerExceptions
+                    .FirstOrDefault(x => x is SpannerException);
+
+            return spannerException != null;
+        }
+
+        private static ErrorCode ConvertFromStatusCode(StatusCode statusCode)
```

Comment:
Just cast? (As per other enums - you might even want an enum converter class to make the validation more efficient)